From 5765b54e8590aea70bbd51652e96e8d8293a08bd Mon Sep 17 00:00:00 2001
From: John Reck <jreck@google.com>
Date: Fri, 12 Jul 2024 09:20:37 +0000
Subject: [PATCH 2/2] Apply pending transactions always

Test: repro in bug
Bug: 334901521

Change-Id: I8f856e7be5917a1d01f0ee2bdbe613bf37045646
(cherry picked from commit 54526e9b29556ede6c22a07fad83cc6cc8163915)
---
 core/java/android/view/ViewRootImpl.java | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/core/java/android/view/ViewRootImpl.java b/core/java/android/view/ViewRootImpl.java
index cd2d36c60ade..f47932fd32c3 100644
--- a/core/java/android/view/ViewRootImpl.java
+++ b/core/java/android/view/ViewRootImpl.java
@@ -4024,7 +4024,15 @@ public final class ViewRootImpl implements ViewParent,
             mReportNextDraw = false;
             mLastReportNextDrawReason = null;
             mActiveSurfaceSyncGroup = null;
-            mSyncBuffer = false;
+            if (mHasPendingTransactions) {
+                // TODO: We shouldn't ever actually hit this, it means mPendingTransaction wasn't
+                // merged with a sync group or BLASTBufferQueue before making it to this point
+                // But better a one or two frame flicker than steady-state broken from dropping
+                // whatever is in this transaction
+                mPendingTransaction.apply();
+                mHasPendingTransactions = false;
+            }            
+	    mSyncBuffer = false;
             if (isInWMSRequestedSync()) {
                 mWmsRequestSyncGroup.markSyncReady();
                 mWmsRequestSyncGroup = null;
-- 
2.34.1

